src/solvers/smt/smt_casts.cpp

```diff
@@ -345,6 +345,60 @@ smt_astt smt_convt::convert_typecast_to_ints_from_bool(const typecast2t &cast)
   return mk_ite(a, one, zero);
 }
 
+static bool can_carry_provenance(const type2tc &t)
+{
+  return t->get_width() >= config.ansi_c.capability_width();
+}
+
+static type2tc capability_struct_type2()
+{
+  assert(
+    config.ansi_c.cheri_concentrate &&
+    "uncompressed CHERI capabilities are not implemented");
+  type2tc type = ptraddr_type2();
+  std::vector<type2tc> members = {type, type};
+  std::vector<irep_idt> names = {"pesbt", "cursor"};
+  return struct_type2tc(members, names, names, "__ESBMC_capability_struct");
+}
+
+static type2tc capability_union_type2()
+{
+  type2tc type = ptraddr_type2();
+  std::vector<type2tc> members = {
+    get_uint_type(config.ansi_c.capability_width()),
+    capability_struct_type2(),
+  };
+  std::vector<irep_idt> names = {"cap", "str"};
+  return union_type2tc(members, names, names, "__ESBMC_capability_union");
+}
+
+static expr2tc capability_struct2(const expr2tc &pesbt, const expr2tc &cursor)
+{
+  std::vector<expr2tc> members = {pesbt, cursor};
+  return constant_struct2tc(capability_struct_type2(), members);
+}
+
+static expr2tc capability_struct_from_cap(const expr2tc &cap)
+{
+  // assert(is_pointer_type(cap));
```

Comment:
Do you know if we need to keep this comment code here?

Reply:
I've changed it to a CHERI-TODO comment for me.